pennylane_lightning/lightning_qubit/lightning_qubit2.py

```diff
@@ -0,0 +1,366 @@
+    validate_observables,
+    no_sampling,
+)
+from pennylane.tape import QuantumTape, QuantumScript
+from pennylane.transforms.core import TransformProgram
+from pennylane.typing import Result, ResultBatch
+
+from ._adjoint_jacobian import LightningAdjointJacobian
+from ._state_vector import LightningStateVector
+from ._measurements import LightningMeasurements
+
+try:
+    # pylint: disable=import-error, unused-import
+    import pennylane_lightning.lightning_qubit_ops
+
+    LQ_CPP_BINARY_AVAILABLE = True
+except ImportError:
+    LQ_CPP_BINARY_AVAILABLE = False
+
+Result_or_ResultBatch = Union[Result, ResultBatch]
+QuantumTapeBatch = Sequence[QuantumTape]
+QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+PostprocessingFn = Callable[[ResultBatch], Result_or_ResultBatch]
+
+
+def simulate(circuit: QuantumScript, state: LightningStateVector) -> Result:
+    """Simulate a single quantum script.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        state (LightningStateVector): handle to Lightning state vector
+
+    Returns:
+        Tuple[TensorLike]: The results of the simulation
+
+    Note that this function can return measurements for non-commuting observables simultaneously.
+    """
+    state.reset_state()
+    final_state = state.get_final_state(circuit)
+    return LightningMeasurements(final_state).measure_final_state(circuit)
+
+
+def jacobian(circuit: QuantumTape, state: LightningStateVector, batch_obs=False):
+    """Compute the Jacobian for a single quantum script.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        state (LightningStateVector): handle to Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
+            computing the jacobian. This value is only relevant when the lightning
+            qubit is built with OpenMP.
+
+    Returns:
+        TensorLike: The Jacobian of the quantum script
+    """
+    state.reset_state()
+    final_state = state.get_final_state(circuit)
+    return LightningAdjointJacobian(final_state, batch_obs=batch_obs).calculate_jacobian(circuit)
+
+
+def simulate_and_jacobian(circuit: QuantumTape, state: LightningStateVector, batch_obs=False):
+    """Simulate a single quantum script and compute its Jacobian.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        state (LightningStateVector): handle to Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
+            computing the jacobian. This value is only relevant when the lightning
+            qubit is built with OpenMP.
+
+    Returns:
+        Tuple[TensorLike]: The results of the simulation and the calculated Jacobian
+
+    Note that this function can return measurements for non-commuting observables simultaneously.
+    """
+    return simulate(circuit, state), jacobian(circuit, state, batch_obs=batch_obs)
```

Comment:
```suggestion
    state.reset_state()
    final_state = state.get_final_state(circuit)
    measurements = LightningMeasurements(final_state).measure_final_state(circuit)
    jacobian = LightningAdjointJacobian(final_state, batch_obs=batch_obs).calculate_jacobian(circuit)
    return [measurements, jacobian]
```

Reply:
This way we just prepare the state one time and can get all the information we need from it.